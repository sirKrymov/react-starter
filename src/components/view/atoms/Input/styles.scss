@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.input {
  min-height: inherit;
  position: relative;

  input {
    -webkit-appearance: none;
    background-color: var(--bgPrimary);
    border-radius: $border-radius;
    transition: border-color $anim-time-fast, background-color $anim-time-fast,
      opacity $anim-time;
    box-shadow: none;
    min-height: inherit;
    padding: 0 16px;
    outline: none;
    border: 1px solid var(--borderPrimary);
    color: var(--textPrimary);
    width: 100%;

    &::placeholder {
      color: var(--textHint);
    }

    &:focus {
      border-color: var(--thirdy);
      box-shadow: none;
    }
  }

  &--right-block-xs {
    input {
      padding-right: 28px;
    }
  }

  &--right-block-sm {
    input {
      padding-right: 36px;
    }
  }

  &--right-block-md {
    input {
      padding-right: 44px;
    }
  }

  &--right-block-lg {
    input {
      padding-right: 52px;
    }
  }

  &--error {
    input {
      border-color: var(--error);
    }
  }

  &__right-block {
    transform: translateY(-50%);
    position: absolute;
    height: 100%;
    right: 0;
    top: 50%;
    @include center;

    &--xs {
      width: 28px;
    }

    &--sm {
      width: 36px;
    }

    &--md {
      width: 44px;
    }

    &--lg {
      width: 52px;
    }
  }
}
