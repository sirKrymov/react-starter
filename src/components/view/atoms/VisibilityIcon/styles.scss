.visibility-icon {
  display: inline-flex;
  cursor: pointer;
}
