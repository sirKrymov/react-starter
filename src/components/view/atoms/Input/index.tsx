// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ChangeEvent, FocusEvent, ReactNode, ReactElement } from 'react';

import cn from 'classnames';

import { UiSize } from 'components/view/atoms/UiSize';

import './styles.scss';

export interface IProps {
  placeholder?: string;
  rightBlock?: ReactNode;
  className?: string;
  onChange(e: ChangeEvent<any> | string): void;
  onFocus?(e: FocusEvent<any>): void;
  onBlur?(e: FocusEvent<any>): void;
  height?: 'xs' | 'sm' | 'md' | 'lg' | 'full';
  width?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full' | 'content';
  error?: boolean;
  type?: 'password' | 'text';
  value: string;
  name: string;
  id: string;
}

export function Input({
  placeholder,
  rightBlock,
  className,
  onChange,
  onFocus,
  onBlur,
  height = 'md',
  width,
  error,
  value,
  name,
  type = 'text',
  id,
  ...rest
}: IProps): ReactElement {
  const inputClass = cn(
    {
      input: true,
      [`input--right-block-${height}`]: rightBlock && height,
      'input--error': error
    },
    className
  );
  const rightBlockClass = cn({
    'input__right-block': true,
    [`input__right-block--${height}`]: rightBlock && height
  });

  return (
    <UiSize width={width} height={height}>
      <div className={inputClass}>
        <input
          autoComplete={type === 'password' ? 'new-password' : 'off'}
          placeholder={placeholder}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          value={value}
          name={name}
          type={type}
          id={id}
          {...rest}
        />

        {rightBlock && <div className={rightBlockClass}>{rightBlock}</div>}
      </div>
    </UiSize>
  );
}
