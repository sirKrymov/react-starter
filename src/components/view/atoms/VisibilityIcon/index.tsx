// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactElement } from 'react';

import { SvgIcon } from '../SvgIcon';

import { eyeCrossedIcon, eyeIcon } from 'constants/icons';

import './styles.scss';

interface IProps {
  isVisible: boolean;
  onClick(): void;
  id: string;
}

export function VisibilityIcon({
  isVisible,
  onClick,
  id
}: IProps): ReactElement {
  return (
    <div className="visibility-icon" onClick={onClick} id={id}>
      <SvgIcon size="lg" src={isVisible ? eyeCrossedIcon : eyeIcon} />
    </div>
  );
}
