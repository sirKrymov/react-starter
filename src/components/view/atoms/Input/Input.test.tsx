import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { VisibilityIcon } from '../VisibilityIcon';
import { Input } from '../Input';

describe('<Input> atom component', () => {
  const handleChange = jest.fn();
  const handleClick = sinon.spy();

  let shallowComponent: ShallowWrapper;

  const createComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    shallowComponent = shallow(
      <Input
        onChange={onChange}
        value={value}
        name="input"
        id="check-input"
        {...rest}
      />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent({ onChange: handleChange, value: '' });
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });

    it('THEN it should fires on change correctly', () => {
      const event = { target: { value: 'input text' } };

      expect(shallowComponent.find('input').simulate('change', event));
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenLastCalledWith(event);
    });
  });

  describe('WHEN the component have different props', () => {
    it('WHEN to component have rightContent', () => {
      createComponent({
        rightContent: (
          <VisibilityIcon
            onClick={handleClick}
            isVisible={false}
            id="password-visibility-icon"
          />
        )
      });

      expect(shallowComponent.find(VisibilityIcon)).toBeDefined();
    });
  });
});
