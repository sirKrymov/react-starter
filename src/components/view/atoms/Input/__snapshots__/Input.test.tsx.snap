// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<UiSize
  height="md"
>
  <div
    className="input"
  >
    <input
      autoComplete="off"
      id="check-input"
      name="input"
      onChange={[MockFunction]}
      type="text"
      value=""
    />
  </div>
</UiSize>
`;
